test/bson_older_versions_tests.js

```diff
@@ -73,18 +76,160 @@ describe('Current version', function () {
+    {
+      name: 'Browser ESM',
+      // eval because import is a syntax error in earlier Node.js versions
+      // that are still supported in CI
+      load: () => eval(`import("${__dirname}/../bson.browser.esm.mjs")`),
+      usesBufferPolyfill: true
+    },
+    {
+      name: 'Browser UMD',
+      load: () => Promise.resolve(require('../dist/bson.browser.umd.js')),
+      usesBufferPolyfill: true
+    },
+    {
+      name: 'Generic bundle',
+      load: () => {
+        const source = fs.readFileSync(__dirname + '/../dist/bson.bundle.js', 'utf8');
+        return Promise.resolve(vm.runInNewContext(`${source}; BSON`, { global, process }));
+      },
+      usesBufferPolyfill: true
+    },
+    {
+      name: 'Node.js ESM',
+      load: () => eval(`import("${__dirname}/../bson.esm.mjs")`)
+    }
+  ]);
+
+  const makeObjects = bson => [
+    new bson.ObjectId('5f16b8bebe434dc98cdfc9ca'),
+    new bson.DBRef('a', new bson.ObjectId('5f16b8bebe434dc98cdfc9cb'), 'db'),
+    new bson.MinKey(),
+    new bson.MaxKey(),
+    new bson.Timestamp(1, 100),
+    new bson.Code('abc'),
+    bson.Decimal128.fromString('1'),
+    bson.Long.fromString('1'),
+    new bson.Binary(Buffer.from('abcäbc🎉'), 128),
+    new Date('2021-05-04T15:49:33.000Z'),
+    /match/
+  ];
+
+  for (const from of variants) {
+    for (const to of variants) {
+      describe(`serializing objects from ${from.name} using ${to.name}`, () => {
+        let fromObjects;
+        let fromBSON;
+        let toBSON;
+
+        before(function () {
+          // Load the from/to BSON versions asynchronously because e.g. ESM
+          // requires asynchronous loading.
+          return Promise.resolve()
+            .then(() => {
+              return from.load();
+            })
+            .then(loaded => {
+              fromBSON = loaded;
+              return to.load();
+            })
+            .then(loaded => {
+              toBSON = loaded;
+            })
+            .then(
+              () => {
+                fromObjects = makeObjects(fromBSON);
+              },
+              err => {
+                if (+process.version.slice(1).split('.')[0] >= 12 && process.platform !== 'win32') {
+                  throw err; // On Node.js 12+, all loading is expected to work.
+                } else {
+                  this.skip(); // Otherwise, e.g. ESM can't be loaded, so just skip.
+                }
+              }
+            );
+        });
+
+        it('serializes in a compatible way', function () {
+          for (const object of fromObjects) {
+            // If the object in question uses Buffers in its serialization, and
+            // its Buffer was created using the polyfill, and we're serializing
+            // using a legacy version that uses buf.copy(), then that fails
+            // because the Buffer polyfill's typechecking is buggy, so we
+            // skip these cases.
+            // This is tracked as https://jira.mongodb.org/browse/NODE-2848
```

Comment:
Just to cement my understanding we could change bson 1.x to use .set and .subarray to correct the incompatibility right?

Reply:
Yes, that would work :+1: